crates/kubelet/src/volumes.rs

```diff
@@ -0,0 +1,186 @@
+    fn as_ref(&self) -> &PathBuf {
+        &self.host_path
+    }
+}
+
+impl Deref for VolumeRef {
+    type Target = PathBuf;
+
+    fn deref(&self) -> &Self::Target {
+        &self.host_path
+    }
+}
+
+impl Drop for VolumeRef {
+    fn drop(&mut self) {
+        if matches!(self.volume_type, VolumeType::ConfigMap | VolumeType::Secret) {
+            // TODO: Currently there is no way to do this async (though there is an async destructors proposal)
+            debug!(
+                "deleting {:?} directory {:?}",
+                self.volume_type, self.host_path
+            );
+            std::fs::remove_dir_all(&self.host_path).unwrap_or_else(|e| {
+                error!(
+                    "unable to delete directory {:?} on volume cleanup: {:?}",
+                    self.host_path, e
+                )
+            });
+        }
+    }
+}
+
+/// This is a gnarly function to check all of the supported data members of the Volume struct.
+/// Because it isn't a HashMap, we need to check all fields individually
+async fn configure(
+    vol: &KubeVolume,
+    namespace: &str,
+    client: &kube::Client,
+    path: &PathBuf,
+) -> anyhow::Result<VolumeType> {
+    if let Some(cm) = &vol.config_map {
+        populate_from_config_map(
+            &cm.name
+                .as_ref()
+                .ok_or_else(|| anyhow::anyhow!("no configmap name was given"))?,
+            namespace,
+            client,
+            path,
+        )
+        .await
+    } else if let Some(s) = &vol.secret {
+        populate_from_secret(
+            &s.secret_name
+                .as_ref()
+                .ok_or_else(|| anyhow::anyhow!("no configmap name was given"))?,
+            namespace,
+            client,
+            path,
+        )
+        .await
+    } else {
+        Err(anyhow::anyhow!(
+            "Unsupported volume type. Currently supported types: ConfigMap, Secret, and HostPath"
+        ))
+    }
+}
+
+async fn populate_from_secret(
+    name: &str,
+    namespace: &str,
+    client: &kube::Client,
+    path: &PathBuf,
+) -> anyhow::Result<VolumeType> {
+    tokio::fs::create_dir(path).await?;
+    let secret_client: Api<Secret> = Api::namespaced(client.clone(), namespace);
+    let secret = secret_client.get(name).await?;
+    let data = secret.data.unwrap_or_default();
+    let data = data.iter().map(|(key, data)| async move {
+        let file_path = path.join(key);
+        tokio::fs::write(file_path, &data.0).await
+    });
+    futures::future::join_all(data)
+        .await
+        .into_iter()
+        .collect::<tokio::io::Result<_>>()?;
+
+    Ok(VolumeType::Secret)
+}
+
+async fn populate_from_config_map(
+    name: &str,
+    namespace: &str,
+    client: &kube::Client,
+    path: &PathBuf,
+) -> anyhow::Result<VolumeType> {
+    tokio::fs::create_dir(path).await?;
+    let cm_client: Api<ConfigMap> = Api::namespaced(client.clone(), namespace);
+    let cm = cm_client.get(name).await?;
```

Comment:
nit: not a fan of the two letter variable name here. Unless the usage is super local (within a line or two of the declaration) I try to avoid naming things so tersely.

Reply:
`config_map` and `config_map_client` felt really verbose. Generally I agree, but I felt like there was enough context here. I can change it for clarity though